python/metrics/src/hydrotools/metrics/_validation.py

```diff
@@ -100,3 +103,65 @@ def raise_for_inconsistent_shapes(
                 array_shape_1=x.shape,
                 array_shape_2=y.shape
                 )
+
+def validate_boolean_categorical_series(
+    series: pd.Series
+    ) -> None:
+    """
+    Validate that series is a boolean categorical variable.
+
+    Parameters
+    ----------
+    series: pandas.Series, required
+        Series to validate.
+
+    Warnings
+    --------
+    UserWarning:
+        Warns if series is not a boolean Categorical and attempts to convert.
+
+    Raises
+    ------
+    pandas.errors.UnsupportedFunctionCall:
+        Raises if series is not a pandas.Series or if series does not contain two categories: True and False
+
+    Returns
+    -------
+    Validated boolean categorical series.
+    """
+    # Check for Series
+    if not isinstance(series, pd.Series):
+        raise pd.errors.UnsupportedFunctionCall(f"{series} is not a pandas.Series")
+
+    # Check for categorical
+    if not hasattr(series, "cat"):
+        message = f"{series} is not a categorical pandas.Series, attempting to convert"
+        warnings.warn(message=message, category=UserWarning)
+        series = series.astype("category")
+
+    # Check for True category
+    no_true_category = True
+    for c in series.cat.categories:
+        if c is True:
+            no_true_category = False
+    if no_true_category:
+        message = f"{series} has no True category, attempting to add"
+        warnings.warn(message=message, category=UserWarning)
+        series = series.cat.add_categories([True])
+
+    # Check for False category
+    no_false_category = True
+    for c in series.cat.categories:
+        if c is False:
+            no_false_category = False
+    if no_false_category:
+        message = f"{series} has no False category, attempting to add"
+        warnings.warn(message=message, category=UserWarning)
+        series = series.cat.add_categories([False])
+
+    # Check number of categories
```

Comment:
Move this check above the type/inclusion checks...no point in testing what each one is if there are too many...

Reply:
There's more to the story.

Here's a use-case example:
```python
# Say we have a flooding threshold we use to transform a continuous 
#  variable into a categorical variable
flood_threshold = 40.0

# Here's our continues variable (volumetric discharge)
flow_values = pd.Series([41.0, 42.0, 43.0])

# Here's where we convert the flow values into categorical values
observed_flood = (flow_values >= flood_threshold).astype("category")
```

`observed_flood` is a categorical `Series` with values `[True, True, True]` and categories `[True]`. This causes an error when attempting to cross-tabulate when generating contingency tables because it does not contain a `False` category. I was trying to implement a user friendly way to:

1. Add `True/False` to the available categories if missing
2. Make sure that if `True` or `False` were added as categories that there are only two categories

For example, if a categorical `Series` has categories `[False, "foo"]`, this would pass because it has two categories. Then `True` would get added resulting in categories `[False, "foo", True]`. This would raise an error down the road. That's why I check for the number of categories last.

Reply:
Refactor to force it to create a new Categorical Series.